File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>(); 
        Map<Character,Character> hashMap = Map.of('(',')', '[', ']', '{', '}'); 

        for(char c : s.toCharArray()){
            if(hashMap.containsKey(c)){
                stack.push(c); 
            } else{
                if(stack.isEmpty()) return false; 

                char top = stack.peek(); 
                if(c == hashMap.get(top)){
                    stack.pop(); 
                } else {
                    return false; 
                }
            }
        }


        return stack.isEmpty(); 
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
from collections import defaultdict 
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        hashMap = {')' : '(', ']' : '[', '}' : '{'}

        for char in s:

            if char in hashMap:
                if stack and stack[-1] == hashMap[char]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(char) 

        return not stack
                